cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh

```diff
@@ -0,0 +1,302 @@
+ * duplicate points, mergeable segments and mergeable point on segments.
+ */
+template <typename index_t,
+          typename T,
+          typename MultiLinestringRange1,
+          typename MultiLinestringRange2>
+std::pair<linestring_intersection_intermediates<vec_2d<T>, index_t>,
+          linestring_intersection_intermediates<segment<T>, index_t>>
+pairwise_linestring_intersection_with_duplicates(MultiLinestringRange1 multilinestrings1,
+                                                 MultiLinestringRange2 multilinestrings2,
+                                                 rmm::mr::device_memory_resource* mr,
+                                                 rmm::cuda_stream_view stream)
+{
+  static_assert(std::is_integral_v<index_t>, "Index type must be integral.");
+  static_assert(std::is_floating_point_v<T>, "Coordinate type must be floating point.");
+
+  auto const num_pairs = multilinestrings1.size();
+  // Compute the upper bound of spaces required to store intersection results.
+  rmm::device_uvector<index_t> num_points_per_pair(num_pairs, stream);
+  rmm::device_uvector<index_t> num_segments_per_pair(num_pairs, stream);
+
+  thrust::uninitialized_fill_n(rmm::exec_policy(stream), num_points_per_pair.begin(), num_pairs, 0);
+  thrust::uninitialized_fill_n(
+    rmm::exec_policy(stream), num_segments_per_pair.begin(), num_pairs, 0);
+
+  detail::pairwise_linestring_intersection_upper_bound_count(multilinestrings1,
+                                                             multilinestrings2,
+                                                             num_points_per_pair.begin(),
+                                                             num_segments_per_pair.begin(),
+                                                             stream);
+
+  // Initialize the intermediates structure
+  auto num_points = thrust::reduce(
+    rmm::exec_policy(stream), num_points_per_pair.begin(), num_points_per_pair.end());
+  auto num_segments = thrust::reduce(
+    rmm::exec_policy(stream), num_segments_per_pair.begin(), num_segments_per_pair.end());
+
+  detail::linestring_intersection_intermediates<vec_2d<T>, index_t> points(
+    num_pairs, num_points, num_points_per_pair, stream, mr);
+  detail::linestring_intersection_intermediates<segment<T>, index_t> segments(
+    num_pairs, num_segments, num_segments_per_pair, stream, mr);
+
+  // Allocate a temporary vector for each thread to keep track of the number of results
+  // from the current multilinestring pair has written.
+  rmm::device_uvector<index_t> num_points_stored_temp(num_pairs, stream);
+  rmm::device_uvector<index_t> num_segments_stored_temp(num_pairs, stream);
+
+  thrust::uninitialized_fill_n(
+    rmm::exec_policy(stream), num_points_stored_temp.begin(), num_pairs, 0);
+  thrust::uninitialized_fill_n(
+    rmm::exec_policy(stream), num_segments_stored_temp.begin(), num_pairs, 0);
```

Comment:
Why isn't `thrust::unintialized_fill_n` using `cudaMemsetAsync` under the hood?

Reply:
By saying policy, do you mean stream here?

Reply:
I guess because not everything you might fill is just an array of bytes. And you can't use `cudaMemsetAsync` to fill anything that is not the same value in every *byte*. E.g. if you wanted to fill an array of ints with the value 1, then you can't use `cudaMemsetAsync`.

Reply:
Oh I see. That's why memset is more efficient for zero-setting since every byte is the same.